go/internal/lifecycle/manager.go

```diff
@@ -0,0 +1,61 @@
+package lifecycle
+
+import (
+	"context"
+	"sync"
+)
+
+type State int
+
+type Manager struct {
+	condState    *sync.Cond
+	currentState State
+}
+
+func NewManager(initialState State) *Manager {
+	return &Manager{
+		currentState: initialState,
+		condState:    sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (m *Manager) UpdateState(state State) {
+	m.condState.L.Lock()
+	if m.currentState != state {
+		m.currentState = state
+		m.condState.Broadcast()
+	}
+	m.condState.L.Unlock()
+}
+
+func (m *Manager) WaitForStateChange(ctx context.Context, state State) (State, error) {
```

Comment:
can you add a comment or rename the state parameter to understand how it works

my first feeling was that it will wait the state to be the one expected, but from what I can see, if you wait for the current state, it will block until the value change?